From 5504fafa5d27c4aae31b416e5f42419df9dc0f5c Mon Sep 17 00:00:00 2001
From: Rajkumar Manoharan <rmanohar@qca.qualcomm.com>
Date: Wed, 30 Nov 2011 14:15:52 +0530
Subject: [PATCH] UPSTREAM: ath9k: Reconfigure tx power on regulatory update

Whenever the regulatory got updated by country IE for the world
roaming cards, need to reconfigure the tx power immediately to
increase the power level.

Signed-off-by: Rajkumar Manoharan <rmanohar@qca.qualcomm.com>

BUG=chrome-os-partner:4866
TEST=check txpower bb registers with AP that beacons power caps in the CC IE

Change-Id: Ibfa8f45918b9ec0b6d62c38de650c8ef0f88a903
---

diff --git a/drivers/net/wireless/ath/ath9k/init.c b/drivers/net/wireless/ath/ath9k/init.c
index f08123e..9081cf8 100644
--- a/drivers/net/wireless/ath/ath9k/init.c
+++ b/drivers/net/wireless/ath/ath9k/init.c
@@ -284,9 +284,27 @@
 {
 	struct ieee80211_hw *hw = wiphy_to_ieee80211_hw(wiphy);
 	struct ath_softc *sc = hw->priv;
-	struct ath_regulatory *reg = ath9k_hw_regulatory(sc->sc_ah);
+	struct ath_hw *ah = sc->sc_ah;
+	struct ath_regulatory *reg = ath9k_hw_regulatory(ah);
+	int ret;
 
-	return ath_reg_notifier_apply(wiphy, request, reg);
+	ret = ath_reg_notifier_apply(wiphy, request, reg);
+
+	if (!ath_is_world_regd(reg) ||
+	    (request->initiator != NL80211_REGDOM_SET_BY_COUNTRY_IE))
+		return ret;
+
+	/* Set tx power */
+	if (ah->curchan)
+		hw->conf.power_level = ah->curchan->chan->max_power;
+
+	sc->config.txpowlimit = 2 * hw->conf.power_level;
+	ath9k_ps_wakeup(sc);
+	ath9k_cmn_update_txpow(ah, sc->curtxpow, sc->config.txpowlimit,
+			       &sc->curtxpow);
+	ath9k_ps_restore(sc);
+
+	return ret;
 }
 
 /*
